pkg/params/validators.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2022-2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package params
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"unsafe"
+)
+
+type TypeHint string
+
+const (
+	TypeString TypeHint = "string"
+	TypeInt    TypeHint = "int"
+	TypeUint   TypeHint = "uint"
+	TypeBool   TypeHint = "bool"
+)
+
+var typeHintValidators = map[TypeHint]ParamValidator{
+	TypeInt:  ValidateNumber, // TODO: more specific
+	TypeUint: ValidateNumber, // TODO: more specific
+	TypeBool: ValidateBool,
+}
+
+type ParamValidator func(value string) error
+
+func ValidateBool(value string) error {
+	value = strings.ToLower(value)
+	if value != "true" && value != "false" {
+		return fmt.Errorf("expected 'true' or 'false'")
+	}
+	return nil
+}
+
+func ValidateNumber(value string) error {
+	_, err := strconv.Atoi(value)
+	if err != nil {
+		return fmt.Errorf("expected numeric value")
+	}
+	return nil
+}
+
+func ValidateNumberRange(min, max int64) func(value string) error {
+	return func(value string) error {
+		number, err := strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			return fmt.Errorf("expected numeric value")
+		}
+		if number < min || number > max {
+			return fmt.Errorf("number out of range: got %d, expected min %d, max %d", number, min, max)
+		}
+		return nil
+	}
+}
+
+func ValidateSlice(validator ParamValidator) func(value string) error {
+	return func(value string) error {
+		for i, val := range strings.Split(value, ",") {
+			if err := validator(val); err != nil {
+				return fmt.Errorf("entry #%d (%q): %w", i+1, val, err)
+			}
+		}
+		return nil
+	}
+}
+
+func ParamAsIntSlice[T int | int8 | int16 | int32 | int64](p *Param, target *[]T) {
+	StringAsIntSlice(p.value, target)
+}
+
+func StringAsIntSlice[T int | int8 | int16 | int32 | int64](s string, target *[]T) {
+	if s == "" {
+		*target = []T{}
+		return
+	}
+	in := strings.Split(s, ",")
+	out := make([]T, 0, len(in))
+	for _, entry := range in {
+		n, _ := strconv.ParseInt(entry, 10, int(unsafe.Sizeof(*new(T))*8))
+		out = append(out, T(n))
+	}
+	*target = out
+}
+
+func ParamAsUintSlice[T uint | uint8 | uint16 | uint32 | uint64](p *Param, target *[]T) {
+	StringAsUintSlice(p.value, target)
+}
```

Comment:
What about making all the `ParamAs*` functions methods of the `Param` type?

It seems to me it's clearer to write `p.AsInt(&foo)` than `params.ParamAsInt(p, &foo)`. 
(strings package is an [exception](https://stackoverflow.com/questions/28023566/why-are-string-functions-not-defined-on-the-string-type) as it's a built-in type, I don't think we have the same concerns here)

Reply:
I tried that, but sadly generic methods aren't allowed - so it would result in one function for each type. Would you prefer that?

Reply:
I kinda like these functions as the cast is hidden from the user, however just returning the value could make it more clear: 

```golang
tracer.Config.MaxRows = pm[gadgets.ParamMaxRows].AsInt()

tracer.config.MinLatency = pm[ParamMinLatency].AsUint()

tracer.Config.AllFiles = pm[types.AllFilesParam].AsBool()
```

I think those are straightforward to read. The first time I read `params.StringAsInt(pm[gadgets.ParamMaxRows], &tracer.config.MaxRows)` I thought that it was using a pointer because it was creating a sort of binding to fill the parameter later on...

What do you think? /cc @blanquicet  @eiffel-fl

Reply:
> I think those are straightforward to read. The first time I read params.StringAsInt(pm[gadgets.ParamMaxRows], &tracer.config.MaxRows) I thought that it was using a pointer because it was creating a sort of binding to fill the parameter later on...

Hah, true - that's common in the flag stuff which is similar to this.

I'm fine with your example (and also think it's easier to understand), but we'd need As(U)Int8/16/32/64() and so on as well - or would you always go for 64bit and let the gadget developer cast to whatever is appropriate?

Reply:
mmm, not sure. I believe we need to implement all options in the library, otherwise what would be the advantage over using `strvconv` directly? (Also I won't worry too much about this point, we can iterate over it in the future as it's isolated)

Reply:
> what would be the advantage over using strvconv directly?
Mainly not feeling obligated to care about errors at that point (because it's already validated - well, as long as you extract the value in the way you validated it, hehe), thus making the code more straightforward.

> Also I won't worry too much about this point, we can iterate over it in the future as it's isolated
I think we should clarify this in this PR (or at least not wait for too long afterwards), as this basically becomes a part of our "API"/best practices for developing gadgets.

Reply:
ok, so I'd suggest to go with `As{Int,Int8,---}` and avoid the casting on the user. (It's a lot of functions to implement but we need to make it once on the life so that should be fine.

Reply:
Just one question about validation, how are we going to control overflows on integers? Like a number not fitting in an Int8? Is it going to be done in the validation step?

Reply:
> Just one question about validation, how are we going to control overflows on integers? Like a number not fitting in an Int8? Is it going to be done in the validation step?

Ah, good question. Maybe we should add more typeHints with specific validators for all the ints/uints, then.
Would you want to panic if you were to extract a value using the As* functions when it's not matching the typeHint?

Reply:
> Maybe we should add more typeHints with specific validators for all the ints/uints, then

Yes!

> Would you want to panic if you were to extract a value using the As* functions when it's not matching the typeHint?

I don't think we need that. If the validation passed then we can be confident that it's good, can't we?

Reply:
> I don't think we need that. If the validation passed then we can be confident that it's good, can't we?

Yeah, from the validation side it should be - but what if you set typeHint to uint8 and try to get a uint32 for example?

Btw: now that we have to handle every type anyways, another solution could be to again use reflection (this is a cold path) in combination with the pointer - that could result in something like:
```golang
gadgetParams(gadgets.ParamMaxRows).CopyTo(&tracer.Config.MaxRows)
```

I'm fine with both solutions.

Reply:
I personally feel more confident with the standard assignment solution. I think the `CopyTo` hides simple logic and makes the code harder to follow.